# Tensors and Tensor Operations

### Overview
In this video, we explore the fundamentals of PyTorch tensors and key operations used in deep learning workflows.  

### Agenda
- **Creating Tensors** → initialize from lists, arrays, or random values.  
- **Understanding Datatypes** → long, float, double, and their role in computations.  
- **Moving Tensors to GPU** → accelerate operations with `.to(device)`.  
- **Basic Tensor Operations** → element-wise addition, subtraction, multiplication, division.  
- **Advanced Tensor Operations** → matrix multiplication, dot product.  
- **Transpose** → reshaping dimensions for compatibility.  
- **Slope Approximation** → derivative intuition using finite differences.  
- **Tensor Functions** → built-in utilities like `sum`, `mean`, `max`, etc.  
- **Reshaping / Slicing / Indexing** → modifying and accessing tensor data.  
- **`vstack` & `hstack` Manipulations** → vertical and horizontal concatenation.  
- **Common Activation Functions** → ReLU, Sigmoid, Tanh in tensor form.  

### Key Takeaway
By the end, you’ll be comfortable creating, manipulating, and performing essential operations on tensors — the building blocks of neural networks.  

# Tensors

A tensor in PyTorch is a multi-dimensional array (like a NumPy array) that can store data such as numbers, vectors, matrices, or higher-dimensional data.

We start by importing **PyTorch**. Different types of tensors are then created to demonstrate basic initialisation methods.  

- **`a`**: A one-dimensional array created directly from a Python list `[1, 2, 3, 4]`.  
- **`v`**: A one-dimensional array from a list of floating-point numbers `[4.6, 5.0, 6.3]`.  
- **`b`**: A 2×3 matrix filled entirely with 0s, created using `tf.zeros`.  
- **`c`**: A 2×3 matrix filled entirely with 1s, created using `tf.ones`.  
- **`d`**: A sequence of integers from 0 to 11 generated with `tf.range`, then reshaped into a 3×4 matrix using `tf.reshape`.
- **`e`**: An array created using NumPy array.
- Converting a PyTorch tensor to a NumPy array using `.numpy()`.  

Printing each tensor shows how PyTorch represents and stores these structures in memory.

In [4]:
import torch
import numpy as np

numpy_array = np.array([[1,2], [3,4]])

# Creating tensors
a = torch.tensor([1, 2, 3, 4])             # from list
v = torch.tensor([4.6,5.0,6.3])            # from list
b = torch.zeros((2, 3))                    # zeros
c = torch.ones((2, 3))                     # ones
d = torch.arange(0, 12).reshape(3, 4)      # range with reshape
e = torch.from_numpy(numpy_array)
a_numpy = a.numpy()                        # Convert to NumPy
print("a:", a)
print("v:", v)
print("Zeros:\n", b)
print("Ones:\n", c)
print("Arange reshaped:\n", d)
print("From NumPy:\n", e)
print("As NumPy:\n", a_numpy, type(a_numpy))

a: tensor([1, 2, 3, 4])
v: tensor([4.6000, 5.0000, 6.3000])
Zeros:
 tensor([[0., 0., 0.],
        [0., 0., 0.]])
Ones:
 tensor([[1., 1., 1.],
        [1., 1., 1.]])
Arange reshaped:
 tensor([[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]])
From NumPy:
 tensor([[1, 2],
        [3, 4]])
As NumPy:
 [1 2 3 4] <class 'numpy.ndarray'>


# Data Types

In PyTorch, 
- `torch.float32` stores decimal values
- `torch.long` stores 64-bit integers, is the standard choice since `torch.int` is not directly available
- `.long()` converts floats to integers by dropping decimals
- `.float()` converts integers to floats by adding `.0`
- `.double()`it changes the tensor’s dtype to 64-bit floating point (double precision)



In [5]:
print("\n--- Tensor Datatypes ---")
print("Tensor a:", a, "-> dtype:", a.dtype)
print("Tensor v:", v, "-> dtype:", v.dtype)

# Type casting
print("Casted a -> float:", a.float())    # convert integers to float
print("Casted v -> long:", v.long())      # convert integers to long (int64)
print("Casted v -> double:", v.double())   # convert to double (float64)


--- Tensor Datatypes ---
Tensor a: tensor([1, 2, 3, 4]) -> dtype: torch.int64
Tensor v: tensor([4.6000, 5.0000, 6.3000]) -> dtype: torch.float32
Casted a -> float: tensor([1., 2., 3., 4.])
Casted v -> long: tensor([4, 5, 6])
Casted v -> double: tensor([4.6000, 5.0000, 6.3000], dtype=torch.float64)


# Moving Tensor To GPU Device 

Use `.to(device)` to move tensors to CPU or GPU depending on availability.
Moving tensors and models to the **GPU** allows PyTorch to take advantage of the GPU's ability to perform many calculations in parallel.  
This makes both **training** and **inference** much faster compared to running only on the CPU.


In [6]:
# Move tensors to GPU if available
if torch.cuda.is_available():
    device = torch.device("cuda")
    print("GPU is available. Moving tensors to GPU.\n")
    
    a = a.to(device)
    v = v.to(device)
    b = b.to(device)
    c = c.to(device)
    d = d.to(device)
else:
    device = torch.device("cpu")
    print("GPU not available. Using CPU.\n")

GPU is available. Moving tensors to GPU.



# Tensor Basic Operations

## Basic Arithmetic Operations

Tensors in PyTorch support **elementwise arithmetic operations**, meaning the operation is applied to each corresponding element of the tensors. 

- **Addition (`u + v`)**: adds each element of `u` to the corresponding element of `v`.  
- **Subtraction (`u - v`)**: subtracts each element of `v` from the corresponding element of `u`.  
- **Division (`v / u`)**: divides each element of `v` by the corresponding element of `u`.  
- **Multiplication (`u * v`)**: multiplies each element of `u` with the corresponding element of `v`.  

In [29]:
# Let us create two dummy matrix 
u = torch.tensor([[0.1, 0.2], [0.3, 0.4]])
v = torch.tensor([[0.5, 0.6], [0.7, 0.8]])

# Let's apply the basic operations on the matrix
print("Addition:\n", u + v)
print("Subtraction\n", u - v)
print("Division:\n", v / u)
print("Elementwise Multiplication:\n", u * v)

Addition:
 tensor([[0.6000, 0.8000],
        [1.0000, 1.2000]])
Subtraction
 tensor([[-0.4000, -0.4000],
        [-0.4000, -0.4000]])
Division:
 tensor([[5.0000, 3.0000],
        [2.3333, 2.0000]])
Elementwise Multiplication:
 tensor([[0.0500, 0.1200],
        [0.2100, 0.3200]])


## Matrix Multiplication

Each output element is the dot product of a row from `M1` and a column from `M2`, using `torch.matmul()`.
 
If  
$M1 = \begin{bmatrix} 0 & 1 \\ 4 & 5 \end{bmatrix}, \quad  
M2 = \begin{bmatrix} 2 & 0 \\ 1 & 3 \end{bmatrix}$  

Then the product is  

$M1 \times M2 = \begin{bmatrix} 1 & 3 \\ 13 & 15 \end{bmatrix}$



In [30]:
# Let us create two dummy matrix 
M1 = torch.tensor([[0, 1], [4, 5]], dtype=torch.float32)
M2 = torch.tensor([[2, 0], [1, 3]], dtype=torch.float32)

# Let's compute
print("M1\n", M1)
print("M2\n", M2)
print("Matrix Multiplication:\n", torch.matmul(M1, M2))

M1
 tensor([[0., 1.],
        [4., 5.]])
M2
 tensor([[2., 0.],
        [1., 3.]])
Matrix Multiplication:
 tensor([[ 1.,  3.],
        [13., 15.]])


## Dot Product
`torch.dot(u, v)` computes the sum of elementwise products between two vectors. 
<div style="text-align: left;">

$u \cdot v = \sum_{i=1}^{n} u_i v_i$

$u = [1.0, 2.0, 3.0], \quad v = [4.6, 5.0, 6.3]$

$ (1\times 4.6) + (2\times 5.0) + (3\times 6.3) = 4.6 + 10 + 18.9 = 33.5 $

</div>




In [38]:
# Let's create two dummy vectors
u = torch.tensor([1, 2, 3], dtype=torch.float32)
v = torch.tensor([4.6, 5, 6.3], dtype=torch.float32)

# Let's compute
print("Dot product:", torch.dot(u, v))

Dot product: tensor(33.5000)


## Transpose

`M1.T` computes the transpose of the matrix `M1`, where rows become columns and columns become rows.

If $M1 = \begin{bmatrix} 0 & 1 \\ 4 & 5 \end{bmatrix}$

Then the transpose is  

$M1^T = \begin{bmatrix} 0 & 4 \\ 1 & 5 \end{bmatrix}$


In [33]:
# Let us create a dummy matrix 
M1 = torch.tensor([[0, 1], [4, 5]], dtype=torch.float32)

# Let's compute
print("Transpose:\n", M1.T)

Transpose:
 tensor([[0., 4.],
        [1., 5.]])


## Slope Approximation (First-order Difference)

`torch.diff(v)` computes the successive differences between elements of tensor `v`.  
Formula:  
<div style="text-align:left">
$\Delta v_i = v_{i+1} - v_i$

$v = [4.6, 5.0, 6.3]$

$[5.0 - 4.6,\; 6.3 - 5.0] = [0.4,\; 1.3]$
</div>


In [34]:
# "Slope-like" difference (successive differences)

# Let us create two dummy matrix 
v = torch.tensor([4.6, 5, 6.3], dtype=torch.float32)

# Let's compute
slope = torch.diff(v)
print("Tensor:", v)
print("Slope approximation:", slope)

Tensor: tensor([4.6000, 5.0000, 6.3000])
Slope approximation: tensor([0.4000, 1.3000])


# Tensor Functions
 
Here we perform basic reduction operations on the tensor:  
- `mean()` computes the average value.  
- `sum()` adds all elements together.  
- `max()` returns the largest value.  
- `argmax()` gives the index of the largest value.

`.item()` - converts the output tensor to single value

In [35]:
a = torch.tensor([1, 2, 4, 3])

print("Tensor a:", a)
print("Mean of a:", a.float().mean().item())   # convert to float for decimal mean
print("Sum of a:", a.sum().item())
print("Max of a:", a.max().item())
print("Argmax of a:", a.argmax().item())

Tensor a: tensor([1, 2, 4, 3])
Mean of a: 2.5
Sum of a: 10
Max of a: 4
Argmax of a: 2


## Reshaping / Slicing / Indexing 

### Reshaping a Tensor in PyTorch

The `view` method is used to change the shape of a tensor without modifying its underlying data.  

In this case, a one-dimensional tensor with 4 elements is reshaped into a **2×2 matrix**.  
The total number of elements remains the same (4), but the arrangement of values is changed from a single row into a two-dimensional grid.

In [40]:
# Reshape 1D tensor a into 2x2
a = torch.tensor([1, 2, 4, 3])

#Let's reshape it
a_reshaped = a.view(2, 2)
print("Tensor a reshaped (2x2):\n", a_reshaped)

Tensor a reshaped (2x2):
 tensor([[1, 2],
        [4, 3]])


### Accessing Elements with Indexing
Indexing is used to access a specific row or a single element:
- `d[0]` → returns the **first row** of the tensor.  
- `d[2, 3]` → returns the **element at row 2, column 3** (indexing starts at 0).  

In [44]:
#Let's create a simple 2D Matrix
d = torch.tensor([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]])

# Display the original 2D tensor d
print("Tensor d:\n", d,"\n")
# Access first row of d
print("First row of d:\n", d[0],"\n")

# Access a single element at row 2, column 3
print("Element [2,3] of d:\n", d[2, 3])

Tensor d:
 tensor([[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]]) 

First row of d:
 tensor([1, 2, 3, 4]) 

Element [2,3] of d:
 tensor(12)


### Slicing in a Tensor

Slicing is used to select ranges of rows or columns from a tensor:

- `d[:, 0]` → selects the **first column** (all rows `:` and only column `0`).  
- `d[1:, 1:3]` → selects a **submatrix** starting from row 1 to the end, and columns 1 up to (but not including) 3.  

This allows flexible extraction of parts of the tensor, such as entire columns or smaller submatrices.

In [47]:
#Let's create a simple 2D Matrix
d = torch.tensor([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]])

# Access first column of d
print("First column of d:\n", d[:, 0],"\n")

# Slice a sub-matrix: rows 1 to end, columns 1 to 2
print("Sub-matrix of d (rows 1:, cols 1:3):\n", d[1:, 1:3])

First column of d:
 tensor([1, 5, 9]) 

Sub-matrix of d (rows 1:, cols 1:3):
 tensor([[ 6,  7],
        [10, 11]])


# Manipulations
 
This code demonstrates stacking tensors — `torch.vstack` stacks them vertically (row-wise), while `torch.hstack` stacks them horizontally (column-wise).  


In [48]:
# Let's create two dummy vectors
u = torch.tensor([1, 2, 3], dtype=torch.float32)
v = torch.tensor([4.6, 5, 6.3], dtype=torch.float32)

# Let's compute
print("\n--- Stacking ---")
print("Vertical stack:\n", torch.vstack((u, v)))
print("Horizontal stack:\n", torch.hstack((u, v)))


--- Stacking ---
Vertical stack:
 tensor([[1.0000, 2.0000, 3.0000],
        [4.6000, 5.0000, 6.3000]])
Horizontal stack:
 tensor([1.0000, 2.0000, 3.0000, 4.6000, 5.0000, 6.3000])


# Common Activation Function

Here we apply common activation functions on tensor `v`
- **Sigmoid:** Maps values into the range (0, 1)
- **Tanh:** Maps values into the range (-1, 1)
- **ReLU:** Replaces negative values with 0  
- **LeakyReLU:** Similar to ReLU but allows a small slope (here 0.1) for negative values instead of making them 0

In [49]:
import torch.nn.functional as F

v = torch.tensor([4.6, 5, 6.3, -2.5], dtype=torch.float32)

# ReLU (max(0, x))
print("ReLU:", F.relu(v))

# Sigmoid (squashes values between 0 and 1)
print("Sigmoid:", torch.sigmoid(v))

# Tanh (squashes values between -1 and 1)
print("Tanh:", torch.tanh(v))

# LeakyReLU (small slope for negatives)
leaky_relu = F.leaky_relu(v, negative_slope=0.1)
print("LeakyReLU:", leaky_relu)

ReLU: tensor([4.6000, 5.0000, 6.3000, 0.0000])
Sigmoid: tensor([0.9900, 0.9933, 0.9982, 0.0759])
Tanh: tensor([ 0.9998,  0.9999,  1.0000, -0.9866])
LeakyReLU: tensor([ 4.6000,  5.0000,  6.3000, -0.2500])


<br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br>

# Parameters in Torch

It’s a **single-layer neuron model** created with `nn.Module`.

- **Weights (`nn.Parameter`)**: A learnable matrix that connects input features to outputs.  
- **Bias (`nn.Parameter`)**: A learnable vector that shifts the output values.  
- **Forward pass**: Performs a linear transformation  
  $
    y = x W^{\!\top} + b
    $

  followed by **ReLU activation** to add non-linearity.  
- **Application to MNIST**:  
    - Each image is flattened into **784 input features**.
    - The model is trained on the training set by adjusting its weights to minimise the loss (difference between predicted and true labels).  
    - The model produces **10 output scores** (one per digit).
    - Using **softmax**, these scores are turned into probabilities.
    - The digit with the highest probability is taken as the **predicted class**.
    - Predictions are compared with the **true labels** from the test set to calculate **accuracy**.
    - The code also prints **true vs predicted labels** for sample test images.
















#### Importing Required Libraries ####

- **`import torch`** → loads the core PyTorch library for creating tensors and running computations on CPU/GPU.  
- **`import torch.nn as nn`** → imports PyTorch’s neural network module, which provides building blocks like layers (`Linear`, `Conv2d`, etc.).  
- **`import torch.nn.functional as F`** → imports a collection of functions (e.g., activation functions like `ReLU`, `softmax`, loss functions) that can be used directly.  
- **`import pickle`** → used to load and save Python objects; in this context, it helps load the MNIST dataset stored in a `.pkl` file.  


In [67]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import pickle

#### Defining a Simple Neuron Model
- `Neuron` is a custom class that inherits from `nn.Module`.  
- **Weights** (`[output_size, input_size]`) and **bias** (`[output_size]`) are defined as learnable parameters.  
- The **forward method** applies a linear transformation followed by ReLU activation.  

This represents a basic neuron layer mapping inputs to outputs with non-linearity.


In [1]:
class Neuron(nn.Module):
    def __init__(self, input_size, output_size):
        super(Neuron, self).__init__()
        
        # Learnable parameters
        self.weight = nn.Parameter(torch.randn(output_size, input_size) * 0.01)
        self.bias = nn.Parameter(torch.zeros(output_size))
    
    def forward(self, x):
        # Linear transformation + ReLU activation
        linear_output = x @ self.weight.T + self.bias
        activated_output = F.relu(linear_output)
        return activated_output

NameError: name 'nn' is not defined

### Loading the MNIST Data

- The file `mnist.pkl` is loaded using `pickle.load`, which provides `(train_set, valid_set, test_set)`.  
- Each set contains `(X, y)` where:  
  - `X` = images (28×28 pixels flattened into **784 features**)  
  - `y` = labels (digits 0–9)  
- These are converted to PyTorch tensors:  
  - Images → `float32`, Labels → `long`  
- In this code, a forward pass is run only on the **test set**.  
- Since the model’s weights are random and no training is done, the predictions are essentially **random guesses**.  


In [2]:
# Load MNIST data
with open("data/mnist.pkl", "rb") as f:
    mnist_data = pickle.load(f, encoding="latin1")

train_set, valid_set, test_set = mnist_data
train_x, train_y = train_set
test_x, test_y = test_set

# Convert to torch tensors
x_test = torch.tensor(test_x, dtype=torch.float32)
y_test = torch.tensor(test_y, dtype=torch.long)

NameError: name 'pickle' is not defined

#### Evaluating the Model on the Test Set
- We run the test data through the model to get output scores.  
- Then apply **softmax** to convert these scores into probabilities.  
- Use **argmax** to select the predicted class for each sample.  
- We compare predictions with the true labels to count how many are correct.  
- Finally, we calculate accuracy as correct predictions divided by total samples.  

In [3]:
# Define model instance
model1 = Neuron(input_size=784, output_size=10)
# Forward pass on the full test set

print(type(model))
output1 = model(x_test)

# Convert scores to probabilities and predictions
probs = torch.softmax(output1, dim=1)
preds = torch.argmax(probs, dim=1)

# Calculate accuracy
correct = (preds == y_test).sum().item()
accuracy = correct / len(y_test)

NameError: name 'Neuron' is not defined

#### Displaying Test Results
- Print the total number of test samples, the count of correct predictions, and the overall accuracy as a percentage.  
- Show a few sample predictions (first 6), displaying both the true label and the predicted label for each image.  


In [4]:
print(f"Total test samples: {len(y_test)}")
print(f"Correct predictions: {correct}")
print(f"Accuracy on test set: {accuracy*100:.2f}%")

# Show a few sample predictions
print("\nSample Predictions (first 6):")
for i in range(6):
    print(f"True: {y_test[i].item()}  Predicted: {preds[i].item()}")

NameError: name 'y_test' is not defined

#### Conclusion
  - A very simple network with only one linear transformation and ReLU.  
  - Takes 784 pixel features as input and produces 10 outputs (one per digit).  
  - With random weights and no training in the code shown, predictions are just random guesses (~13% accuracy).  
  - Useful as a baseline to demonstrate how inputs pass through weights and biases to generate outputs.


# Multi Layer Perceptron

It’s a **multi-layer perceptron model** created with `nn.Module`.

- **First layer (`nn.Linear`)**:  
  - Connects the **784 input features** (28×28 pixels from each MNIST image) to a **hidden layer** of neurons.  
  - Each hidden neuron learns its own weights and bias.  

- **Activation function (`ReLU`)**:  
  - Applied after the first layer.  
  - Keeps only positive values and sets negatives to zero, adding non-linearity.  

- **Second layer (`nn.Linear`)**:  
  - Connects the **hidden layer** to the **10 output classes** (digits 0–9).  
  - Produces one raw score (logit) for each digit.  

- **Forward pass**:  
  - Data flows through the input layer → hidden layer → output layer.  
  - Computation:  

    $h = \text{ReLU}(x W_{1}^{\!\top} + b_{1}), \quad y = h W_{2}^{\!\top} + b_{2}$

- **Application to MNIST**:  
  - Each image is flattened into **784 input features**.  
  - The model outputs **10 scores**, one for each digit.  
  - Using `softmax`, these scores are converted into probabilities.  
  - The digit with the highest probability is taken as the **predicted class**.  
  - Predictions are compared with the **true labels** from the test set to calculate **accuracy**.  
  - The code also prints **true vs predicted labels** for sample test images.

#### Importing Required Libraries ####

- **`import torch`** → loads the core PyTorch library for creating tensors and running computations on CPU/GPU.  
- **`import torch.nn as nn`** → imports PyTorch’s neural network module, which provides building blocks like layers (`Linear`, `Conv2d`, etc.).  
- **`import torch.nn.functional as F`** → imports a collection of functions (e.g., activation functions like `ReLU`, `softmax`, loss functions) that can be used directly.
- **`import torch.optim as optim`** → loads PyTorch’s optimisation module. Provides algorithms like **SGD** and **Adam** to update model weights during training.  
- **`import pickle`** → used to load and save Python objects; in this context, it helps load the MNIST dataset stored in a `.pkl` file.  

In [72]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
import pickle

#### Defining a Multi-Layer Perceptron (MLP)

The `MLP` class inherits from `nn.Module` and defines a simple feedforward network with two fully connected layers:  
- `fc1` maps inputs to a hidden layer and applies ReLU activation.  
- `fc2` maps the hidden layer to the outputs (logits).  

In the forward pass, data flows through `fc1 → ReLU → fc2`, producing the final output scores.

In [73]:
class MLP(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(MLP, self).__init__()
        # Two fully connected layers
        self.fc1 = nn.Linear(input_size, hidden_size)  # input → hidden
        self.fc2 = nn.Linear(hidden_size, output_size) # hidden → output
    
    def forward(self, x):
        x = F.relu(self.fc1(x))   # first layer + ReLU
        x = self.fc2(x)           # second layer (logits)
        return x

#### Loading the MNIST Data
The MNIST dataset is loaded from a pickle file and split into training, validation, and test sets.  
- From this, we use the **training set** (`train_x`, `train_y`) and the **test set** (`test_x`, `test_y`).  
- Images are converted into `float32` tensors (`x_train`, `x_test`) for model input.  
- Labels are converted into `long` tensors (`y_train`, `y_test`) for classification tasks.  


In [74]:
# Load MNIST data
with open("data/mnist.pkl", "rb") as f:
    mnist_data = pickle.load(f, encoding="latin1")

train_set, valid_set, test_set = mnist_data
train_x, train_y = train_set
test_x, test_y = test_set

# Convert to torch tensors
x_train = torch.tensor(train_x, dtype=torch.float32)
y_train = torch.tensor(train_y, dtype=torch.long)
x_test = torch.tensor(test_x, dtype=torch.float32)
y_test = torch.tensor(test_y, dtype=torch.long)

#### Model, Loss, and Optimiser Setup

- We set the model dimensions with an input size of 784 (28×28 pixels), a hidden layer of 128 neurons, and an output size of 10 for the digit classes 0–9. 
- For training, we use **CrossEntropyLoss** to measure classification error and the **SGD optimiser** with a learning rate of 0.1 to update the model’s parameters.  


In [75]:
# Model, loss, optimizer
input_size = 784   # 28x28 pixels
hidden_size = 128  # hidden neurons
output_size = 10   # 10 digit classes

model_mlp = MLP(input_size, hidden_size, output_size)
criterion = nn.CrossEntropyLoss()          # loss function
optimizer = optim.SGD(model_mlp.parameters(), lr=0.1)  # optimizer

### Training Loop

- Train the model for 5 epochs with mini-batches of size 64.  
- For each batch:  
  - Pass inputs (`xb`) through the model to get predictions.
  - Compare predictions with `yb` using the loss function. 
  - Compute the loss using the criterion.  
  - Perform backpropagation: reset gradients, compute new gradients, and update weights with the optimiser.  
- After each epoch, print the loss to monitor training progress.  


In [76]:
# Training loop
epochs = 5
batch_size = 64

for epoch in range(epochs):
    for i in range(0, len(x_train), batch_size):
        xb = x_train[i:i+batch_size]
        yb = y_train[i:i+batch_size]

        # Forward pass
        output = model_mlp(xb)

        # Compute loss
        loss = criterion(output, yb)

        # Backpropagation
        optimizer.zero_grad()
        loss.backward()

        # Update weights
        optimizer.step()
    
    print(f"Epoch {epoch+1}/{epochs}, Loss: {loss.item():.4f}")

Epoch 1/5, Loss: 0.3737
Epoch 2/5, Loss: 0.2253
Epoch 3/5, Loss: 0.1664
Epoch 4/5, Loss: 0.1225
Epoch 5/5, Loss: 0.0858


### Evaluation on the Test Set

- Switching to evaluation mode with `torch.no_grad()` to disable gradient tracking.  
- Passing the test data (`x_test`) through the model to get output scores.  
- Applying **softmax** and use **argmax** to select the predicted class for each sample.  
- Comparing predictions with true labels (`y_test`) to count correct matches and calculate accuracy.  
- Printing the final test accuracy as a percentage, and also showing a few sample predictions (first 10) with their true labels to see how the model performs on individual cases.  


In [77]:
# Evaluation on test set
with torch.no_grad():
    output = model_mlp(x_test)
    preds = torch.argmax(F.softmax(output, dim=1), dim=1)

    correct = (preds == y_test).sum().item()
    accuracy = correct / len(y_test)

print(f"\nTest Accuracy: {accuracy*100:.2f}%")

# Show some sample predictions
print("\nSample Predictions (first 10):")
for i in range(10):
    print(f"True: {y_test[i].item()}  Predicted: {preds[i].item()}")


Test Accuracy: 96.23%

Sample Predictions (first 10):
True: 7  Predicted: 7
True: 2  Predicted: 2
True: 1  Predicted: 1
True: 0  Predicted: 0
True: 4  Predicted: 4
True: 1  Predicted: 1
True: 4  Predicted: 4
True: 9  Predicted: 9
True: 5  Predicted: 6
True: 9  Predicted: 9


#### Conclusion
  - Extends the Neuron into a two-layer network with a hidden layer and ReLU activation.  
  - Capable of learning more complex patterns when trained.  
  - In practice, after training, it achieves much higher accuracy on MNIST (96%).  
  - Demonstrates how adding depth and non-linearity improves performance.


# Regularisation and Normalisation in Torch

- **Dropout (`nn.Dropout`)**: Randomly switches off neurons during training to prevent overfitting.  
- **Batch Normalisation (`nn.BatchNorm1d`)**: Normalises activations, speeding up and stabilising training.  
- **L2 Regularisation (`weight_decay` in `optim.SGD`)**: Adds a penalty on large weights to keep them small.  
- **Early Stopping (manual check on validation loss)**: Stops training if validation loss does not improve for a set number of epochs.  

#### Application to MNIST
These techniques help the MLP generalise better, avoid memorising training data, and achieve higher accuracy on the test set.

#### Importing Required Libraries ####

- **`import torch`** → loads the core PyTorch library for creating tensors and running computations on CPU/GPU.  
- **`import torch.nn as nn`** → imports PyTorch’s neural network module, which provides building blocks like layers (`Linear`, `Conv2d`, etc.).  
- **`import torch.nn.functional as F`** → imports a collection of functions (e.g., activation functions like `ReLU`, `softmax`, loss functions) that can be used directly.
- **`import torch.optim as optim`** → loads PyTorch’s optimisation module. Provides algorithms like **SGD** and **Adam** to update model weights during training.  
- **`import pickle`** → used to load and save Python objects; in this context, it helps load the MNIST dataset stored in a `.pkl` file.  

In [97]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
import pickle

#### MLP with Dropout and Batch Normalisation
This `MLP` class defines a feedforward neural network with two fully connected layers and additional techniques to improve training:  

- **fc1**: first fully connected layer that maps the input to a hidden layer.  
- **BatchNorm (bn1)**: normalises activations in the hidden layer to stabilise and speed up learning.  
- **Dropout**: randomly drops some hidden units during training to reduce overfitting.  
- **fc2**: second fully connected layer that maps the hidden layer to the final outputs (logits).  

**Forward pass**: the input passes through `fc1`, then batch normalisation and ReLU activation, followed by dropout, and finally through `fc2` to produce the output scores.  

In [98]:
# Define a Multi-Layer Perceptron (MLP) model
class MLP1(nn.Module):
    def __init__(self, input_size, hidden_size, output_size, dropout_rate=0.5):
        super(MLP1, self).__init__()
        self.fc1 = nn.Linear(input_size, hidden_size)
        self.bn1 = nn.BatchNorm1d(hidden_size)   # Normalisation
        self.dropout = nn.Dropout(dropout_rate) # Dropout
        self.fc2 = nn.Linear(hidden_size, output_size)
    
    def forward(self, x):
        x = F.relu(self.bn1(self.fc1(x)))   # input → hidden
        x = self.dropout(x)                 # dropout
        x = self.fc2(x)                     # hidden → output
        return x


### Loading the MNIST Data

The MNIST dataset is loaded from a pickle file and split into three parts: **training**, **validation**, and **test** sets.  
- Each set contains images (`x`) and labels (`y`).  
- Images (`train_x`, `valid_x`, `test_x`) are converted into `float32` tensors for model input.  
- Labels (`train_y`, `valid_y`, `test_y`) are converted into `long` tensors, which is the required type for classification tasks in PyTorch.  

This gives us separate datasets for training the model, validating performance during training, and testing on unseen data.  

In [99]:
# Load MNIST data
with open("data/mnist.pkl", "rb") as f:
    mnist_data = pickle.load(f, encoding="latin1")

train_set, valid_set, test_set = mnist_data
train_x, train_y = train_set
valid_x, valid_y = valid_set
test_x, test_y = test_set

# Convert to torch tensors
x_train = torch.tensor(train_x, dtype=torch.float32)
y_train = torch.tensor(train_y, dtype=torch.long)
x_valid = torch.tensor(valid_x, dtype=torch.float32)
y_valid = torch.tensor(valid_y, dtype=torch.long)
x_test = torch.tensor(test_x, dtype=torch.float32)
y_test = torch.tensor(test_y, dtype=torch.long)

#### Model, Loss, and Optimiser Setup
- The model is defined with an input size of 784 (flattened 28×28 images), a hidden layer of 128 neurons, and an output size of 10 classes for digits 0–9.  
- The `MLP` is created with a dropout rate of 0.3.  
- **CrossEntropyLoss** is used as the loss function for multi-class classification.  
- The **SGD optimiser** is configured with a learning rate of 0.1 and **L2 regularisation** (`weight_decay=1e-4`) to penalise large weights and reduce overfitting.  


In [100]:
# Model, loss, optimizer
input_size = 784
hidden_size = 128
output_size = 10

model_reg = MLP1(input_size, hidden_size, output_size, dropout_rate=0.3)
criterion = nn.CrossEntropyLoss()

# Using L2 regularisation via weight_decay
optimizer = optim.SGD(model_reg.parameters(), lr=0.1, weight_decay=1e-4)

### Training Loop with Early Stopping

- Training for 30 epochs with mini-batches of size 64.  
- Each batch: forward pass, compute loss, backpropagate, and update weights.  
- After each epoch: evaluating on the validation set, calculate loss and accuracy, and print results.  
- Early stopping: if validation loss improves, save the model; if not improved for 3 epochs, stop and restore the best model.  


In [101]:
# Training loop with Early Stopping
epochs, batch_size, patience = 30, 64, 3
best_val_loss, no_improve_epochs = float('inf'), 0

for epoch in range(epochs):
    model_reg.train()
    for i in range(0, len(x_train), batch_size):
        xb, yb = x_train[i:i+batch_size], y_train[i:i+batch_size]
        loss = criterion(model_reg(xb), yb)
        optimizer.zero_grad(); loss.backward(); optimizer.step()

    model_reg.eval()
    with torch.no_grad():
        val_output = model_reg(x_valid)
        val_loss = criterion(val_output, y_valid)
        val_preds = torch.argmax(F.softmax(val_output, dim=1), dim=1)
        val_acc = (val_preds == y_valid).float().mean().item()

    print(f"Epoch {epoch+1}/{epochs}, Train Loss: {loss.item():.4f}, "
          f"Val Loss: {val_loss.item():.4f}, Val Acc: {val_acc:.4f}")
   
    if val_loss.item() < best_val_loss:  # Early stopping
        best_val_loss, no_improve_epochs = val_loss.item(), 0
        best_model_state = model_reg.state_dict()
    else:
        no_improve_epochs += 1
        if no_improve_epochs >= patience:
            print("Early stopping triggered."); break

model_reg.load_state_dict(best_model_state)

Epoch 1/30, Train Loss: 0.3600, Val Loss: 0.1752, Val Acc: 0.9498
Epoch 2/30, Train Loss: 0.0997, Val Loss: 0.1386, Val Acc: 0.9604
Epoch 3/30, Train Loss: 0.0638, Val Loss: 0.1237, Val Acc: 0.9653
Epoch 4/30, Train Loss: 0.1808, Val Loss: 0.1116, Val Acc: 0.9686
Epoch 5/30, Train Loss: 0.1136, Val Loss: 0.1039, Val Acc: 0.9699
Epoch 6/30, Train Loss: 0.1011, Val Loss: 0.0967, Val Acc: 0.9725
Epoch 7/30, Train Loss: 0.1724, Val Loss: 0.0910, Val Acc: 0.9743
Epoch 8/30, Train Loss: 0.0758, Val Loss: 0.0911, Val Acc: 0.9729
Epoch 9/30, Train Loss: 0.0415, Val Loss: 0.0907, Val Acc: 0.9728
Epoch 10/30, Train Loss: 0.0191, Val Loss: 0.0863, Val Acc: 0.9754
Epoch 11/30, Train Loss: 0.0477, Val Loss: 0.0847, Val Acc: 0.9756
Epoch 12/30, Train Loss: 0.0319, Val Loss: 0.0828, Val Acc: 0.9758
Epoch 13/30, Train Loss: 0.0408, Val Loss: 0.0836, Val Acc: 0.9751
Epoch 14/30, Train Loss: 0.0154, Val Loss: 0.0821, Val Acc: 0.9776
Epoch 15/30, Train Loss: 0.0671, Val Loss: 0.0814, Val Acc: 0.9773
Epoc

<All keys matched successfully>

### Evaluation on the Test Set

- Switching the model to evaluation mode with `model.eval()` and disable gradient tracking using `torch.no_grad()`.  
- Running the test data through the model to get output scores.  
- Applying **softmax** to convert scores into probabilities and use **argmax** to pick the predicted class for each sample.  
- Comparing predictions with true labels (`y_test`) to count correct matches and compute overall accuracy.  
- Printing the final test accuracy as a percentage, and also showing a few sample predictions (first 10) with their true labels to see how the model performs on individual cases. 


In [103]:
# Evaluation on test set
model_reg.eval()
with torch.no_grad():
    test_output = model_reg(x_test)
    preds = torch.argmax(F.softmax(test_output, dim=1), dim=1)

    correct = (preds == y_test).sum().item()
    accuracy = correct / len(y_test)

print(f"\nTest Accuracy: {accuracy*100:.2f}%")

print("\nSample Predictions (first 10):")
for i in range(10):
    print(f"True: {y_test[i].item()}  Predicted: {preds[i].item()}")


Test Accuracy: 97.64%

Sample Predictions (first 10):
True: 7  Predicted: 7
True: 2  Predicted: 2
True: 1  Predicted: 1
True: 0  Predicted: 0
True: 4  Predicted: 4
True: 1  Predicted: 1
True: 4  Predicted: 4
True: 9  Predicted: 9
True: 5  Predicted: 6
True: 9  Predicted: 9


#### Conclusion
  - Builds on MLP by adding Batch Normalisation, Dropout, and L2 regularisation.  
  - Includes early stopping based on validation loss for better generalisation.  
  - These techniques prevent overfitting and stabilise training.  
  - Provides the most balanced and robust results among the three with 98% accuracy.